include/boost/compat/shared_lock.hpp

```diff
@@ -0,0 +1,160 @@
+// Copyright 2023 Christian Mazakas.
+// Distributed under the Boost Software License, Version 1.0.
+// https://www.boost.org/LICENSE_1_0.txt
+
+#ifndef BOOST_COMPAT_SHARED_LOCK_HPP_INCLUDED
+#define BOOST_COMPAT_SHARED_LOCK_HPP_INCLUDED
+
+#include <boost/config.hpp>
+#include <boost/throw_exception.hpp>
+
+#include <memory>
+#include <mutex>
+#include <system_error>
+
+namespace boost {
+namespace compat {
+
+template <class Mutex>
+class shared_lock;
+
+template <class Mutex>
+void swap( shared_lock<Mutex>& x, shared_lock<Mutex>& y ) noexcept;
+
+namespace detail {
+
+BOOST_NORETURN BOOST_NOINLINE inline void
+throw_system_error( std::errc e, boost::source_location const& loc ) {
```

Comment:
You can add a default argument, `loc = BOOST_CURRENT_LOCATION`, to avoid having to pass it each time.